### PR TITLE
prop_conv_solvert: introduce scopes

### PR DESCRIPTION
**Pull request by kroening**
This is the biggest change to the solver interface since it got added to the codebase.

Currently, bound identifiers (exists, forall, let) must be unique; the solver looses soundness if that assumption is violated.

This change introduces scopes, i.e., the uniqueness requirement is dropped. Nested scoping works as expected. This matches what SMT-LIB does, i.e., the renaming done in the SMT-LIB front-end can be dropped.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — COMMENTED** [review]
Looks OK. CI is failing, though.

**Comment by peterschrammel** [comment]
Tests for prop_conv and smt2 would be required.

File: src/solvers/flattening/boolbv.h [hunk 1]
```diff
@@ -119,6 +119,10 @@ class boolbvt:public arrayst
   typedef std::unordered_map<const exprt, bvt, irep_hash> bv_cachet;
   bv_cachet bv_cache;
 
+  // record the subexpressions that use the scoped variable,
+  // in order to clear them from the cache after we are done
+  std::vector<bv_cachet::iterator> bv_cache_iterators;
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv.h]: Iterators into an unordered map aren't stable. Is there a guarantee that we will only ever put iterators in here after the source map isn't modified anymore?

File: src/solvers/flattening/boolbv_let.cpp [hunk 2]
```diff
@@ -15,16 +15,30 @@ bvt boolbvt::convert_let(const let_exprt &expr)
...
+  bool previous_record_cache_iterators = record_cache_iterators;
+  std::size_t previous_prop_cache_iterators_size = prop_cache_iterators.size();
+  std::size_t previous_bv_cache_iterators_size = bv_cache_iterators.size();
+  record_cache_iterators = true;
+
+  // recursive call
   bvt result_bv=convert_bv(expr.where());
 
+  // clear cache_iterators that use the bound symbol from the cache
+  bv_cache_iterators.resize(previous_prop_cache_iterators_size);
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_let.cpp]: `prop_cache_iterators.resize`?

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/solvers/flattening/boolbv.h (hunk 1 above)
Comment (kroening) [reply src/solvers/flattening/boolbv.h]: Uh, indeed.
This now makes me worried about `boolbvt::convert_bv`. It assumes that `cache_result.first` remains valid during `convert_bitvector`, which I don't believe is the case.

Comment (tautschnig) [reply src/solvers/flattening/boolbv.h]: Making a few wild guesses: what makes iterators unstable is a possible `rehash` happening. Quite possibly the iterator object itself is ok, as is any reference/pointer contained in there to a value. What will have changed, however, is the relative position of objects, and thus going to the next or previous entry would not work. If this is true, maybe my concerns for this code are unfounded?

**Comment by kroening** [comment]
This will need #4931 to work.

**Comment by kroening** [comment]
This needs #4992.

File: src/solvers/flattening/boolbv.h [hunk 3]
```diff
@@ -271,6 +271,9 @@ class boolbvt:public arrayst
 
   // strings
   numberingt<irep_idt> string_numbering;
+
+  // scopes
+  std::size_t scope_counter = 0;
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv.h:276]: Random place for comment:
- In the commit message fix s/looses/loses/
- The commit message promises "(which is the next  commit)" -- but there is no such change to be found.

**Review by tautschnig — COMMENTED** [review]
1) See admin comments about what's being promised vs what is delivered below.
2) Could `array_comprehension_exprt` please also make use of scopes?

File: src/solvers/flattening/boolbv.h (hunk 3 above)
Comment (kroening) [reply src/solvers/flattening/boolbv.h:276]: Commit message fixed.

**Comment by kroening** [comment]
> 2. Could `array_comprehension_exprt` please also make use of scopes?

There are many more expressions with scope (forall and exists in particular). Will deliver!

**Review by tautschnig — APPROVED** [review]

**Pull request merged**